File: Atcoder_#300/javascript/c_cross.js
function main(input) {
    // 入力の処理
    const lines = input.trim().split('\n');
    const [h, w] = lines[0].split(' ').map(Number);
    const grid = lines.slice(1).map(line => line.split(''));

    const n = Math.min(h, w); // N は min(H, W)

    // 各サイズのバツ印の数を格納する配列
    const counts = Array(n).fill(0);

    // バツ印の中心を探索
    for (let i = 1; i < h - 1; i++) {
        for (let j = 1; j < w - 1; j++) {
            if (grid[i][j] !== '#') continue;

            let maxSize = 0;
            
            // バツ印のサイズごとにチェック
            for (let size = 1; size < n; size++) {
                if (i - size < 0 || i + size >= h || j - size < 0 || j + size >= w) break;

                if (grid[i + size][j + size] === '#' &&
                    grid[i + size][j - size] === '#' &&
                    grid[i - size][j + size] === '#' &&
                    grid[i - size][j - size] === '#') {
                    maxSize = size;
                } else {
                    break;
                }
            }

            // 最大サイズのバツ印をカウント
            if (maxSize > 0) {
                counts[maxSize - 1]++;
            }
        }
    }

    // 結果を出力
    console.log(counts.join(' '));
}

// 標準入力からデータを取得して main 関数を呼び出す
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin', 'utf8');
main(input);

File: Atcoder_#300/ruby/c_cross.rb
def main
  # 入力
  h, w = gets.split.map(&:to_i)
  grid = Array.new(h) { gets.chomp.chars }

  # N は min(H, W)
  n = [h, w].min

  # 各サイズのバツ印の数を格納する配列
  counts = Array.new(n, 0)

  # バツ印の中心を探索
  (1...h-1).each do |i|
    (1...w-1).each do |j|
      next unless grid[i][j] == '#'

      # バツ印のサイズごとにチェック
      max_size = 0

      (1...n).each do |size|
        # 中心からの距離 size の4つのマスを確認
        if i - size < 0 || i + size >= h || j - size < 0 || j + size >= w
          break
        end

        if grid[i + size][j + size] == '#' &&
           grid[i + size][j - size] == '#' &&
           grid[i - size][j + size] == '#' &&
           grid[i - size][j - size] == '#'
          max_size = size
        else
          break
        end
      end

      # 最大サイズのバツ印をカウント
      counts[max_size - 1] += 1 if max_size > 0
    end
  end

  # 結果の出力
  puts counts.join(' ')
end

main
